homeassistant/components/roborock/scene.py

```diff
@@ -0,0 +1,74 @@
+"""Support for Roborock scene."""
+
+from __future__ import annotations
+
+import asyncio
+from typing import Any
+
+from homeassistant.components.scene import Scene as SceneEntity
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import EntityDescription
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import RoborockConfigEntry
+from .coordinator import RoborockDataUpdateCoordinator
+from .entity import RoborockEntity
+
+
+async def _get_device_scenes(
+    coordinator: RoborockDataUpdateCoordinator,
+) -> list[SceneEntity]:
+    return [
+        RoborockSceneEntity(
+            coordinator,
+            EntityDescription(
+                key=str(scene.id),
+                name=scene.name,
+            ),
+        )
+        for scene in await coordinator.get_scenes()
+    ]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: RoborockConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up scene platform."""
+    devices_scene_entities = await asyncio.gather(
+        *[
+            _get_device_scenes(coordinator)
+            for coordinator in config_entry.runtime_data.v1
+        ],
+    )
+    async_add_entities(
+        entity
+        for device_scene_entities in devices_scene_entities
+        for entity in device_scene_entities
+    )
```

Comment:
Something seems off to me creating the entity in another function. I think its better to create entities in `async_setup_entry` and closer to `async_add_entities`. 

I see the reason why you separated this is because the coordinator has multiple scenes, so I would suggest doing it this way to separate the scene fetching from the entity creation:

```suggestion
    scene_lists = await asyncio.gather(
        *(
            coordinator.get_scenes()
            for coordinator in config_entry.runtime_data.v1
        ),
    )

    async_add_entities(
        RoborockSceneEntity(
            coordinator,
            EntityDescription(
                key=str(scene.id),
                name=scene.name,
            ),
        )
        for coordinator, scenes in zip(config_entry.runtime_data.v1, scene_lists)
        for scene in scenes
    )
```

I may have some of the syntax off, but its something like that.

Reply:
done